scripts/plugins/extauth-hook-AD.py

```diff
@@ -9,138 +9,304 @@
-    # and the list of subjects from xapi
-    try:
-        config = PamSshConfig()
-        subjects = session.xenapi.subject.get_all()
-        admin_role = session.xenapi.role.get_by_name_label('pool-admin')[0]
+    def _apply_to_cache(self):
+        # AD is not enabled, just return as the configure file will be removed during install
+        if not self._ad_enabled:
+            return
+        # Rewrite the PAM SSH config using the latest info from Active Directory
+        # and the list of subjects from xapi
+        subjects = self._session.xenapi.subject.get_all()
+        admin_role = self._session.xenapi.role.get_by_name_label('pool-admin')[0]
         # Add each subject which contains the admin role
         for opaque_ref in subjects:
-            subject_rec = session.xenapi.subject.get_record(opaque_ref)
+            subject_rec = self._session.xenapi.subject.get_record(opaque_ref)
             sid = subject_rec['subject_identifier']
+            name = subject_rec["other_config"]["subject-name"]
+            is_group = subject_rec["other_config"]["subject-is-group"] == "true"
+            logger.debug("Permit %s with sid %s is_group as %s", name, sid, is_group)
             if admin_role in subject_rec['roles']:
-                config.add_subject(sid)
-        config.install()
+                self._add_subject(name, is_group)
+
+    def _add_subject(self, name, is_group):
+        condition = "ingroup {}".format(name) if is_group else "= {}".format(name)
+        self._lines.append("account sufficient pam_succeed_if.so user {} ".format(condition))
+
+    def _install(self):
+        if self._ad_enabled:
+            super(DynamicPam, self)._install()
+        else:
+            if os.path.exists(self._file_path):
+                os.remove(self._file_path)
+
+
+class KeyValueConfig(ADConfig):
+    _empty_line_prefix = "__key_vluae_config_empty_line_prefix_" # Presume normal config does not have such keys
+    _empty_value = ""
+
+    def __init__(self, path, session, args, ad_enabled=True, load_existing=True, file_mode=0o644, sep=": ", comment="#"):
+        super(KeyValueConfig, self).__init__(path, session, args, ad_enabled, load_existing, file_mode)
+        self._sep = sep
+        self._comment = comment
+        self._values = OrderedDict()
+        self._load_values()
+
+    def _is_comment_line(self, line):
+        return line.startswith(self._comment)
+
+    def _is_empty_line(self, line):
+        return line.startswith(self._empty_line_prefix)
+
+    def _load_values(self):
+        for idx, line in enumerate(self._lines):
+            if line == "": # Empty line
+                key = "{}{}".format(self._empty_line_prefix, str(idx)) # Generate a unique key to store multiple empty lines
+                self._values[key] = self._empty_value
+            elif self._is_comment_line(line):
+                self._values[line] = self._empty_value # Store the comment as key, Presume comments are unique
```

Comment:
can't index be used here, like on the empty lines?

Reply:
yes, it can be used,  but I think that is over designed, 

The index needs to be append to the comments and split out during save,
The files we update actually does not have duplicated comments, so let's just keep it simple.